### PR TITLE
PR: Fix crash and error dialog when triggering OverflowError (too large int) when editing arrays and dataframes

### LINKED ISSUE #6114: ArrayEditor crashes Spyder, DataFrame Editor generates OverflowError when entering a value for a cell larger than the max int for the current int size
## Description

Upon entering/pasting a value into a cell in a dataframe or array editor opened from Variable Explorer that is large enough to cause an overflow of that type, Spyder doesn't catch the resulting `OverflowError` and handle it appropriately, instead displaying an error dialog when the cell is clicked off to commit the value (which is, of course, left unchanged). Additionally, for arrays, something in the exception handling isn't right, such that this situation crashes the entire program with the standard "Python has stopped working error"...*except* if ``SPYDER_DEBUG=3``, in which case the error message is just printed to the shell spyder is started from and everything else proceeds normally.

**What steps will reproduce the problem?**

1. Create a pandas DataFrame of type ``int32`` or ``int64`` (other integer types aren't supported at all), e.g. with
```
import pandas as pd
df = pd.DataFrame([0, 0, 0], dtype=np.int32)
```
or an array, with e.g.
```
import numpy as np
test_array = np.array([0, 0, 0], dtype=np.int32)
``` 
2. Open the dataframe or array from Variable Explorer in a DataFrame/Array Editor 
3. Paste in a value larger than the data type can hold, e.g. ``17179869184`` for ``int32`` or ``73786976294838206464`` for ``int64``, and hit enter or click off.

**What is the expected output? What do you see instead?**

Expected: No hard crash with arrays, and a popup warning dialog similar to the behavior entering a text string in a float column.

When entering a value too large for the current type (int32) but not larger than 64 bit (int64) in an array:
 
```
Traceback (most recent call last):
  File "C:\Users\christopher.gerlach\Anaconda3\lib\site-packages\spyder\widgets\variableexplorer\arrayeditor.py", line 318, in setData
    self.test_array[0] = val # will raise an Exception eventually
OverflowError: Python int too large to convert to C long
```
 
When entering a value greater than an int64 into an int64 cell in an array:
 
```
Traceback (most recent call last):
  File "C:\Users\christopher.gerlach\Anaconda3\lib\site-packages\spyder\widgets\variableexplorer\arrayeditor.py", line 318, in setData
    self.test_array[0] = val # will raise an Exception eventually
OverflowError: int too big to convert
```
 
Further, both cases then print the following (followed by a hard crash, “Python has stopped working” etc., except in debug mode (SPYDER_DEBUG=3), where everything seems to work fine other than printing the error to the debug console, and of course not actually updating the array with the new value):
 
```
During handling of the above exception, another exception occurred:
 
Traceback (most recent call last):
  File "C:\Users\christopher.gerlach\Anaconda3\lib\site-packages\spyder\widgets\variableexplorer\arrayeditor.py", line 320, in setData
    print(type(e.message))  # spyder: test-skip
AttributeError: 'OverflowError' object has no attribute 'message'
```
 
 
 
For Dataframes, similar occurs but the error is caught by Spyder before it crashes.
 
When entering a value too large for the current type (int32) but not larger than 64 bit (int64) in a df:
 
```
  File "C:\Users\christopher.gerlach\Anaconda3\lib\site-packages\spyder\widgets\variableexplorer\dataframeeditor.py", line 368, in setData
    self.df.iloc[row, column-1] = current_value.__class__(val)
OverflowError: Python int too large to convert to C long
```
 
When entering a value greater than an int64 into an int64 cell in a df:
 
```
File "C:\Users\christopher.gerlach\Anaconda3\lib\site-packages\spyder\widgets\variableexplorer\dataframeeditor.py", line 368, in setData
self.df.iloc[row, column-1] = current_value.__class__(val)
OverflowError: int too big to convert
```

**Please provide any additional information below**

**Proposed Solution:**

First of all, fix the exception handling that allows the application hard crash to occur. Then, catch the error in each case, and generate a popup warning dialog with the error similar to the behavior when entering e.g. a value that cannot be converted to the specified data type, like a text string in a numeric column. Ideally, both this case and the “cannot convert” one mentioned should abort while leaving the user with the edit-mode text field with the value they entered so they can more easily correct it, rather than just kicking them out to non-edit mode, showing the previous value, but this would probably take more work than the current behavior.
 

## Version and main components

* Spyder Version: 3.2.5 
* Python Version: 3.6.3
* Qt Versions: 5.6.2, PyQt5 5.6 on Windows

## Dependencies
```
pyflakes >=0.6.0 :  1.6.0 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pygments >=2.0   :  2.2.0 (OK)
pandas >=0.13.1  :  0.21.1 (OK)
numpy >=1.7      :  1.13.3 (OK)
sphinx >=0.6.6   :  1.6.3 (OK)
rope >=0.9.4     :  0.10.5 (OK)
jedi >=0.9.0     :  0.10.2 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
sympy >=0.7.3    :  1.1.1 (OK)
cython >=0.21    :  0.26.1 (OK)
qtconsole >=4.2.0:  4.3.1 (OK)
IPython >=4.0    :  6.1.0 (OK)
pylint >=0.25    :  1.7.4 (OK)

```

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
Fixes #6114 .

Spyder array and dataframe editor now fail gracefully when a value too large for the given (int) datatype is entered in a cell, showing a warning message rather than crashing or displaying an error dialog, respectively (the former due to a Python3-incompatible ``e.message`` syntax of retrieving the error string in that case). Also, a bit of minor associated cleanup.

TODO: Add a simple test for each, ETA Tomorrow...err, later today (Wednesday). Assuming that happens in time, it could be included in 3.2.6, though no pressure on my end.

Also, one really weird bug I wasn't yet able to track down: Before the Array Editor change, Spyder would crash completely ("Python has stopped working...") due to the exception handling bug, *except* when SPYDER_DEBUG=3. In that case, it would just print the error traceback to the shell it was launched in and continue on. However, conversely, once this fix was implemented it now works fine in all scenarios, *except* when SPYDER_DEBUG=3 and one hits enter/return to confirm the whole Array editor dialog, rather than just clicking off of the current cell to save just it. In the former case, the warning dialog is triggered as it should, but after hitting okay, the Variable Explorer window c;loses and then a "Python has stopped working..." crash is triggered. I was able to get the following traceback printed to console:

```
QObject::installEventFilter(): Cannot filter events for objects in a different thread.
Traceback (most recent call last):
  File "C:\Users\C. A. M. Gerlach\Documents\dev\SpyderDev\spyder\spyder\widgets\variableexplorer\arrayeditor.py", line 393, in commitAndCloseEditor
    self.closeEditor.emit(editor, QAbstractItemDelegate.NoHint)
RuntimeError: wrapped C/C++ object of type QLineEdit has been deleted
```

When I tried catching and passing that error, it would still crash, presumably somewhere else maybe in C++ that I couldn't see. Again, this only happened in SPYDER_DEBUG=3, when hitting enter right from the edit box, and with an OverflowError of course. If you have any insight, please let me know; otherwise just ignore this and I'll look into it more tomorrow...er, today.

### CODE REVIEW

**Comment by CAM-Gerlach** [comment]
All builds on Travis, but not AppVeyor or CircleCI, are failing all four tests, apparently due to self.test_array[0] = val not raising an error (OverflowError) when it should, despite confirming that ``val`` is the same as I'm seeing on my machine, i.e. ``1 ** 34**. I'm baffled as to why that might be; it appears both are running identical or nearly so numpy and python versions, and the difference is the same across Python 3.6, 3.5 and 2.7; only different CIs running the test suite. Any ideas?

**Comment by jitseniesen** [comment]
There is also no OverflowError being raised on my Linux box when assigning `2 ** 34` to an `int32` array. Here is a plain (i.e. outside Spyder) IPython session:

```
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct 13 2017, 12:02:49)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.0.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: arr = np.array([1,2,3], dtype=np.int32)

In [3]: arr[0] = 2 ** 34

In [4]: arr
Out[4]: array([0, 2, 3], dtype=int32)

In [5]: arr[0] = 2 ** 66
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-5-3cf8afbda221> in <module>()
----> 1 arr[0] = 2 ** 66

OverflowError: Python int too large to convert to C long
```

**Comment by jitseniesen** [comment]
numpy/numpy#6289 seems to confirm this

**Comment by CAM-Gerlach** [comment]
@jitseniesen Sorry to bother you again @jitseniesen , but any idea on this one? Its only `test_dataframeeditor_edit_overflow` that's failing, and it does so on every Travis build consistently; the OverflowError is triggered as it should be as the unit test of that passes, and the overall qtbot navigation seems to work also since ``test_arrayeditor_edit_overflow`` passes too. If it isn't immediately obvious, maybe test it with `qtbot.wait(1000)`'s and see what it does, if you can? I really need to get my old Linux machine up and running, but I'm not even sure if its bootable at this point.

File: spyder/widgets/variableexplorer/tests/test_dataframeeditor.py [hunk 1]
```diff
@@ -263,5 +265,72 @@ def test_sort_dataframe_with_category_dtypes(qtbot):  # cf. issue 5361
...
+        assert not model.setData(index, str(int(2 ** bit_exponent)))
+        MockQMessageBox.critical.assert_called_with(ANY, "Error", ANY)
+        assert MockQMessageBox.critical.call_count == idx
+        assert numpy.sum(test_df[0].as_matrix() ==
+                         model.df.as_matrix()) == len(test_df)
+
+
+@flaky(max_runs=3)
+@pytest.mark.skipif(os.environ.get('CI', None) is not None and
+                    system() == "Linux",
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/tests/test_dataframeeditor.py]: Please use `sys.platform.startswith('linux')` here

**Comment by ccordoba12** [comment]
@CAM-Gerlach, I left a minor comment, then I'd say this is ready.

File: spyder/widgets/variableexplorer/tests/test_dataframeeditor.py (hunk 1 above)
Comment (CAM-Gerlach) [reply spyder/widgets/variableexplorer/tests/test_dataframeeditor.py]: Sure, that's what I used on the other one previously, but couldn't remember the exact syntax. Just curious, what's the advantage over platform.system()?

Comment (ccordoba12) [reply spyder/widgets/variableexplorer/tests/test_dataframeeditor.py]: It's just our custom throughout our entire codebase.

Comment (CAM-Gerlach) [reply spyder/widgets/variableexplorer/tests/test_dataframeeditor.py]: Makes sense, thanks.

**Comment by CAM-Gerlach** [comment]
Updated as requested, plus more minor cleanup. CIs are still running, but I'm going to try to get a little sleep before work so wanted to let you know before. Thanks!

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Checks passed, so should be ready to go.

**Pull request merged**